File: src/main/java/exnihiloomnia/items/misc/ItemSilkworm.java
package exnihiloomnia.items.misc;

import exnihiloomnia.blocks.leaves.TileEntityInfestedLeaves;
import exnihiloomnia.items.ENOItems;
import exnihiloomnia.util.helpers.InventoryHelper;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.util.EnumActionResult;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.EnumHand;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;

public class ItemSilkworm extends Item {
	
    public ItemSilkworm() {
        super();
        this.setCreativeTab(ENOItems.ENO_TAB);
    }

    @Override
    public EnumActionResult onItemUse(ItemStack stack, EntityPlayer playerIn, World worldIn, BlockPos pos, EnumHand hand, EnumFacing facing, float hitX, float hitY, float hitZ) {
        if (TileEntityInfestedLeaves.infest(worldIn, pos, true)) {
            InventoryHelper.consumeItem(playerIn, stack);
            return EnumActionResult.SUCCESS;
        }
        
        return EnumActionResult.PASS;
    }
}

File: src/main/kotlin/com/jozufozu/exnihiloomnia/common/items/ItemSilkWorm.kt
package com.jozufozu.exnihiloomnia.common.items

import com.jozufozu.exnihiloomnia.common.blocks.leaves.BlockSilkwormInfested
import com.jozufozu.exnihiloomnia.common.lib.ItemsLib
import net.minecraft.item.ItemUseContext
import net.minecraft.util.ActionResultType
import net.minecraft.util.SoundCategory

class ItemSilkWorm : ModItem(ItemsLib.SILKWORM) {
    override fun onItemUse(p_195939_1_: ItemUseContext): ActionResultType {
        if (!worldIn.isRemote) {
            val uninfestedState = worldIn.getBlockState(pos)

            BlockSilkwormInfested.blockMappings[uninfestedState.block]?.let {
                worldIn.setBlockState(pos, it.uninfestedToInfested(uninfestedState))

                val soundType = uninfestedState.block.getSoundType(uninfestedState, worldIn, pos, player)
                worldIn.playSound(null, pos, soundType.placeSound, SoundCategory.BLOCKS, soundType.getVolume() * 0.3f, 0.3f * worldIn.rand.nextFloat() + 1.5f)

                player.swingArm(hand)

                if (!player.isCreative) {
                    player.getHeldItem(hand).shrink(1)
                }

                return EnumActionResult.SUCCESS
            }
        }

        return super.onItemUse(player, worldIn, pos, hand, facing, hitX, hitY, hitZ)
    }
}